google-cloud-storage/acceptance/storage/bucket_policy_test.rb

```diff
@@ -0,0 +1,105 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "storage_helper"
+
+describe Google::Cloud::Storage::Bucket, :storage do
+  let(:bucket_name) { $bucket_names[1] } # Policy version 1
+  let :bucket do
+    storage.bucket(bucket_name) ||
+    safe_gcs_execute { storage.create_bucket(bucket_name) }
+  end
+  let(:bucket_name_policy_v3) { $bucket_names[2] }
+  let :bucket_policy_v3 do
+    storage.bucket(bucket_name_policy_v3) ||
+      safe_gcs_execute { storage.create_bucket(bucket_name_policy_v3) }
+  end
+
+  describe "Policy version 1" do
+    it "allows policy to be updated" do
+      # Check permissions first
+      roles = ["storage.buckets.getIamPolicy", "storage.buckets.setIamPolicy"]
+      permissions = bucket.test_permissions roles
+      skip "Don't have permissions to get/set bucket's policy" unless permissions == roles
+
+      bucket.policy.must_be_kind_of Google::Cloud::Storage::Policy
+
+      # We need a valid service account in order to update the policy
+      service_account = storage.service.credentials.client.issuer
+      service_account.wont_be :nil?
+      role = "roles/storage.objectCreator"
+      member = "serviceAccount:#{service_account}"
+      bucket.policy do |p|
+        p.add role, member
+        p.add role, member # duplicate member will not be added to request
+      end
+
+      role_member = bucket.policy.role(role).select { |x| x == member }
+      role_member.count.must_equal 1
+    end
+
+    it "allows permissions to be tested" do
+      roles = ["storage.buckets.delete", "storage.buckets.get"]
+      permissions = bucket.test_permissions roles
+      permissions.must_equal roles
+    end
+  end
+
+  describe "Policy version 3" do
+    it "allows policy version to be set to 3" do
+      bucket_policy_v3.uniform_bucket_level_access = true
+      # Check permissions first
+      roles = ["storage.buckets.getIamPolicy", "storage.buckets.setIamPolicy"]
+      permissions = bucket_policy_v3.test_permissions roles
+      skip "Don't have permissions to get/set bucket's policy" unless permissions == roles
+
+      policy = bucket_policy_v3.policy
+      policy.must_be_kind_of Google::Cloud::Storage::Policy
+      policy.version.must_equal 1
+      policy.roles.count.must_equal 2
+      expect { policy.bindings }.must_raise RuntimeError
+
+      # We need a valid service account in order to update the policy
+      service_account = storage.service.credentials.client.issuer
+      service_account.wont_be :nil?
+      role = "roles/storage.admin"
+      member = "serviceAccount:#{service_account}"
+      bucket_policy_v3.policy do |p|
+        p.version = 3
+        p.bindings.push({
```

Comment:
The only concern I have is that the bindings is an array of hash objects, and not an array of Binding objects. An array of hash objects might be sufficient for an apiary client like Storage, but is this design the best choice for a grpc client like Pub/Sub? I would prefer that we expose similar APIs in both clients.

Reply:
@frankyn What do you think about a `Binding` value object? This sort of simple encapsulation is typical for the Ruby client libraries, and among other benefits may ease the transition from REST to gRPC. See [`Bucket::Cors::Rule`](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-storage/lib/google/cloud/storage/bucket/cors.rb#L174-L207) for an example.

Reply:
+1 for binding objects. It's a good practice to make it clear, and easily checkable and documentable, what the fields are.

Reply:
@jkwlui Is it OK to use/provide `Binding` value objects instead of raw hashes?

Reply:
Option 1. With Condition value object:

```rb
p.bindings.push(
  Google::Cloud::Storage::Policy::Binding.new(          
    role: role,
    members: [member],
    condition: Google::Cloud::Storage::Policy::Binding::Condition.new(
      title: "always-true",
      description: "test condition always-true",
      expression: "true"
    )
  )
)
```

Option 2. With flattened condition attrs:

```rb
p.bindings.push(
  Google::Cloud::Storage::Policy::Binding.new(
    role: role,
    members: [member],
    condition_title: "always-true",
    condition_description: "test condition always-true",
    condition_expression: "true"
  )
)
```

Option 3. With flattened condition attrs, and Array delegate`add` helper method:

```rb
p.bindings.add(
  role: role,
  members: [member],
  condition_title: "always-true",
  condition_description: "test condition always-true",
  condition_expression: "true"
)
```

Reply:
Option 4. Array delegate `add` method with condition argument as a Hash.

```ruby
p.bindings.add(
  role: role,
  members: [member],
  condition: {
    title: "always-true",
    description: "test condition always-true",
    expression: "true"
  }
)
```

This approach could also be used for `Binding.new` as well:

```ruby
p.bindings.push(
  Google::Cloud::Storage::Policy::Binding.new(          
    role: role,
    members: [member],
    condition: {
      title: "always-true",
      description: "test condition always-true",
      expression: "true"
    }
  )
)
```

Basically, check if the `condition` is a Hash, and if it is, pass that Hash to the `.new` method as the arguments.

Reply:
I'm inclined _not_ to flatten the condition arg here. Flattening introduces a bit more ambiguity (if, say, a toplevel `condition_expression` were provided but not `condition_title`, whereas if they were within a `condition` subhash, it's a bit more obvious what the intent is.) I like Mike's option, except I'd rather not have both `add` and `push`, as it's not obvious which to use for which case. I think there are too many mutation methods of `Array` to use as our API (don't forget the bang methods of `Enum`). Could we just have `bindings=` which would go through the given array elements and accept `Binding` objects and convert hashes?

Reply:
Performing a conversion on `#bindings=` seems like a good idea, and I do think we should support that.  But it seems to be a more likely use case is to add or remove a single binding than to overwrite all the existing bindings. I think there still is a place for a helper method on the `bindings` array for easy creation of a new binding. For instance I would rather type this:

```ruby
p.bindings.add(...)
```

Thank have to type this:

```ruby
p.bindings.push Google::Cloud::Storage::Policy::Binding.new(...)
```

I do concede that `#add` might not be the best name. Perhaps something like `#add_binding` instead?

Reply:
> Could we just have bindings= which would go through the given array elements and accept Binding objects and convert hashes?

A `Policy#bindings=` method would **replace** the existing array with a new array, correct? If so, how would you add entries to the existing array?

Reply:
Per offline conversation, we are going to explore replacing the `bindings` array with a custom `Bindings` object that manages the `Binding` and `Condition` value objects in an encapsulated array. I'll post example code soon.

Reply:
Here is a simple spike on creating a `Bindings` object that holds all the Binding objects. It accepts either Hash or Binding objects, allowing them to be inserted or removed in bulk, but does not expose a mutable Array object. This helps us avoid non-valid values from being inserted into the array, which is a huge problem with the existing `Policy#roles` implementation.

```ruby
class Condition
  def initialize title:, description:, expression:
    @title = title
    @description = description
    @expression = expression
  end
end

class Binding
  def initialize role:, members:, condition: nil
    @role = String role

    @members = Array members
    raise ArgumentError, "" if @members.empty?

    condition = Condition.new **condition if condition.is_a? Hash
    raise ArgumentError, "" unless condition.is_a? Condition
    @condition = condition
  end
end

class Bindings
  include Enumerable
  
  def initialize
    @bindings = []
  end

  def insert *bindings
    bindings = coerce_bindings *bindings
    @bindings += bindings
  end

  def remove *bindings
    bindings = coerce_bindings *bindings
    @bindings -= bindings
  end

  def each
    return enum_for :each unless block_given?

    @binding.each { |binding| yield binding }
  end

  protected
  
  def coerce_bindings *bindings
    bindings.map do |binding|
      binding = Binding.new **binding if binding.is_a? Hash
      raise ArgumentError, "" unless binding.is_a? Binding
      binding
    end
  end
end

def Policy
  attr_reader :bindings
  
  def initialize
    @bindings = Bindings.new
  end

  def binding role:, members:, condition: nil
    Binding.new role: role, members: members, condition: condition
  end
end
```

Reply:
@blowmage Thanks for this proposal! I think we should explore it in a branch PR, just in case I need to make further (unrelated) changes to this PR.

Reply:
draft pr for this  at #4477.

Reply:
merged